### PR TITLE
Add scalar version of transpose conv s8

### PR DESCRIPTION
**Pull request by mansnils**
Adds initial support for transpose conv with unit tests as well.

### CODE REVIEW

File: Include/arm_nnfunctions.h [hunk 1]
```diff
@@ -21,7 +21,7 @@
  * Title:        arm_nnfunctions.h
  * Description:  Public header file for CMSIS NN Library
  *
- * $Date:        10 October 2023
+ * $Date:        27 October 2023
  * $Revision:    V.12.1.0
```
Comment (felix-johnny) [inline Include/arm_nnfunctions.h:25]: Please update the version number.

File: Include/arm_nnfunctions.h [hunk 2]
```diff
@@ -355,11 +355,97 @@ arm_cmsis_nn_status arm_convolve_s8(const cmsis_nn_context *ctx,
...
+                                          const cmsis_nn_dims *output_dims,
+                                          int8_t *output_data);
+
+/**
+ * @brief Get the required buffer size for s8 transpose conv function
+ *
+ * @param[in]       input_dims            Input (activation) tensor dimensions. Format: [N, H, W, C_IN]
+ * @param[in]       filter_dims           Filter tensor dimensions. Format: [C_OUT, HK, WK, C_IN] where HK and WK
+ * are the spatial filter dimensions
+ * @param[in]       output_dims           Output tensor dimensions. Format: [N, H, W, C_OUT]
```
Comment (felix-johnny) [inline Include/arm_nnfunctions.h:412]: out_dims instead of output_dims

File: Include/arm_nnsupportfunctions.h [hunk 3]
```diff
@@ -370,6 +370,35 @@ arm_cmsis_nn_status arm_nn_mat_mult_nt_t_s8(const int8_t *lhs,
                                             const int32_t activation_max,
                                             const int32_t lhs_cols_offset);
 
+/**
```
Comment (felix-johnny) [inline Include/arm_nnsupportfunctions.h:373]: Please update the file version and date..

File: Source/ConvolutionFunctions/arm_transpose_conv_s8.c [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+            {
+                memcpy(img_data, bias_data, output_ch * sizeof(int32_t));
+                img_data += output_ch;
+            }
+        }
+
+        int32_t *col_data_ptr = col_data;
+        const int8_t *filter_data_ptr = filter_data;
+
+        memset(col_data_ptr, 0, col_buf_size);
```
Comment (felix-johnny) [inline Source/ConvolutionFunctions/arm_transpose_conv_s8.c:128]: Please use arm_memset_s8 or arm_memset_s8 instead.

File: Source/NNSupportFunctions/arm_nn_mat_mult_nt_t_s32.c [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+ * @{
+ */
+
+/*
+ * s32 matrix multiplication with the right-hand-side matrix transposed
+ *
+ * Refer header file for details.
+ *
+ */
+arm_cmsis_nn_status arm_nn_mat_mult_nt_t_s32(const int8_t *lhs,
```
Comment (felix-johnny) [inline Source/NNSupportFunctions/arm_nn_mat_mult_nt_t_s32.c:48]: The naming convention used here is primarily of the input and output when they are the same.. and if they are different like in arm_nn_lstm_calculate_gate_s8_s16.c, then it is of input followed by the output. Following that, could you rename this for a sufix of s8_s32 instead?

**Review by felix-johnny — COMMENTED** [review]
Nice work! I've added some minor comments..

File: Include/arm_nnfunctions.h (hunk 1 above)
Comment (mansnils) [reply Include/arm_nnfunctions.h]: Done

File: Include/arm_nnfunctions.h (hunk 2 above)
Comment (mansnils) [reply Include/arm_nnfunctions.h]: Done

File: Include/arm_nnsupportfunctions.h (hunk 3 above)
Comment (mansnils) [reply Include/arm_nnsupportfunctions.h:373]: Done

File: Source/ConvolutionFunctions/arm_transpose_conv_s8.c (hunk 4 above)
Comment (mansnils) [reply Source/ConvolutionFunctions/arm_transpose_conv_s8.c]: Done

File: Source/NNSupportFunctions/arm_nn_mat_mult_nt_t_s32.c (hunk 5 above)
Comment (mansnils) [reply Source/NNSupportFunctions/arm_nn_mat_mult_nt_t_s32.c]: Yes done

**Comment by felix-johnny** [comment]
Looks ok. I'll let Adrian do the final ok..

**Review by AdrianLundell — APPROVED** [review]

**Pull request merged**